File: src/rag-module/fusion.ts
type RankedHit = {
  qid: string;
  score: number;
};

export function rrfFuse(rankings: RankedHit[][], rrfK: number): RankedHit[] {
  const merged = new Map<string, number>();

  for (const list of rankings) {
    for (let i = 0; i < list.length; i++) {
      const qid = list[i].qid;
      const add = 1 / (rrfK + i + 1);
      merged.set(qid, (merged.get(qid) ?? 0) + add);
    }
  }

  return Array.from(merged.entries())
    .map(([qid, score]) => ({ qid, score }))
    .sort((a, b) => b.score - a.score);
}

File: rag_python/fusion.py
from __future__ import annotations


def rrf_fuse(rankings: list[list[dict]], rrf_k: int) -> list[dict]:
    merged: dict[str, float] = {}
    for rows in rankings:
        for i, row in enumerate(rows):
            merged[row["qid"]] = merged.get(row["qid"], 0.0) + 1.0 / (rrf_k + i + 1)
    out = [{"qid": qid, "score": score} for qid, score in merged.items()]
    out.sort(key=lambda x: x["score"], reverse=True)
    return out
